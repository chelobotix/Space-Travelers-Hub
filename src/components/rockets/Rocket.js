import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import style from './Rocket.module.css';

// import style from './Mission.module.css'
import {
  joinRocket,
  leaveRocket,
} from '../../redux/features/rockets/rocketsSlice';

const Rocket = (props) => {
  const {
    id, name, description, image, reserved,
  } = props;
  const dispatch = useDispatch();

  const handleReserveRocket = (id) => {
    dispatch(joinRocket(id));
  };

  const handleLeaveReserveRocket = (id) => {
    dispatch(leaveRocket(id));
  };

  return (
    <>
      <div className={`${style.grid}`}>
        <div className={`${style.border}`}>
          <img src={image[0]} alt="rocketImage" className={`${style.img}`} />
        </div>
        <div className={style.descriptionContainer}>
          <h3 data-testid="rocketName">{name}</h3>
          <p className={style.description}>
            {reserved && (
              <span className={style.reserved}>
                Reserved
              </span>
            )}
            {description}
          </p>
          {!reserved ? (
            <button
              type="button"
              className={`${style.btn1}`}
              onClick={() => handleReserveRocket(id)}
            >
              Reserve Rocket
            </button>
          ) : (
            <button type="button" className={`${style.btn2}`} onClick={() => handleLeaveReserveRocket(id)}>
              Cancel Reservation
            </button>
          )}
        </div>
      </div>
    </>
  );
};

Rocket.defaultProps = {
  id: 0, name: '', description: '', image: null, reserved: false,
};
Rocket.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  description: PropTypes.string,
  image: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.number, PropTypes.string])),
  reserved: PropTypes.bool,
};

export default Rocket;
