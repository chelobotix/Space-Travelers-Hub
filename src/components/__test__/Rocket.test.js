import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import Rocket from '../rockets/Rocket';

const rocketsArray = [
  {
    id: 1,
    rocket_name: 'Rocket Falcon 1',
    description: 'Description Test',
    image: ['image1.jpg', 'image2.jpg'],
    reserved: false,
  },
];

const mockStore = configureStore([]);
const initialState = {
  missions: {
    missions: rocketsArray,
  },
};

const store = mockStore(initialState);

describe('Rocket', () => {
  test('Get Rocket Name', () => {
    render(
      <Provider store={store}>
        <Rocket
          id={rocketsArray[0].id}
          name={rocketsArray[0].rocket_name}
          description={rocketsArray[0].description}
          image={rocketsArray[0].image}
          reserved={rocketsArray[0].reserved}
        />
      </Provider>,
    );
    const td = screen.getByTestId('rocketName');
    expect(td).toHaveTextContent('Rocket Falcon 1');
  });

  test('Rocket Tree', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Rocket
          id={rocketsArray[0].id}
          name={rocketsArray[0].rocket_name}
          description={rocketsArray[0].description}
          image={rocketsArray[0].image}
          reserved={rocketsArray[0].reserved}
        />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Click button text', () => {
    render(
      <Provider store={store}>
        <Rocket
          id={rocketsArray[0].id}
          name={rocketsArray[0].rocket_name}
          description={rocketsArray[0].description}
          image={rocketsArray[0].image}
          reserved={rocketsArray[0].reserved}
        />
      </Provider>,
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(button.textContent).toBe('Reserve Rocket');
  });

  test('Click button listener', () => {
    const handleJoinMission = jest.fn();
    handleJoinMission.mockReturnValue(store.missionsArray);
    handleJoinMission();
    render(
      <Provider store={store}>
        <Rocket
          id={rocketsArray[0].id}
          name={rocketsArray[0].rocket_name}
          description={rocketsArray[0].description}
          image={rocketsArray[0].image}
          reserved={rocketsArray[0].reserved}
        />
      </Provider>,
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(handleJoinMission).toHaveBeenCalled();
  });
});
