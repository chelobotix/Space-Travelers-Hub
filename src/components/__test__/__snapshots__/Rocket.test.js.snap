// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission Rocket Tree 1`] = `
<div
  className="grid"
>
  <div
    className="border"
  >
    <img
      alt="rocketImage"
      className="img"
      src="image1.jpg"
    />
  </div>
  <div
    className="descriptionContainer"
  >
    <h3
      data-testid="rocketName"
    >
      Rocket Falcon 1
    </h3>
    <p
      className="description"
    >
      Description Test
    </p>
    <button
      className="btn1"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;

exports[`Rocket Rocket Tree 1`] = `
<div
  className="grid"
>
  <div
    className="border"
  >
    <img
      alt="rocketImage"
      className="img"
      src="image1.jpg"
    />
  </div>
  <div
    className="descriptionContainer"
  >
    <h3
      data-testid="rocketName"
    >
      Rocket Falcon 1
    </h3>
    <p
      className="description"
    >
      Description Test
    </p>
    <button
      className="btn1"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
